geode-core/src/main/java/org/apache/geode/cache/query/internal/index/CompactRangeIndex.java

```diff
@@ -165,7 +165,7 @@ void removeMapping(RegionEntry entry, int opCode) throws IMQException {
       // we know in this specific case, that a before op was called and stored oldKey/value
       // we also know that a regular remove won't work due to the entry no longer being present
       // We know the old key so let's just remove mapping from the old key
-      if (oldKeyValue != null) {
+      if (oldKeyValue != null && oldKeyValue.get() != null) {
```

Comment:
Is there a test case that isn't mocked that shows this problem?  It might be nice to know how the value in oldKeyValue is null.  The concern is that if we hit the other branch of code, the index updates will have a very noticeable decline in performance.  We would be crawling the entire index for every time the oldKeyValue.get() is returning null.

PartitionedRegionCompactRangeIndexDUnitTest has a few tests that were added when this line of code was added.  If we had something similar (even just for testing), we might be able to find a different fix and avoid the crawling of the entire index.

Reply:
Problem can be reproduced, as described in issue:
On REPLICATE_PERSISTENT region with defined indexes, in configuration with 2 servers, continues traffic (put, delete) is run. After one server is restarted, NullPointerException occurres. With UT is simulated scenario observed from logs.